include/singa/core/tensor.h

```diff
@@ -100,11 +104,93 @@ class Tensor {
     return shape_.at(idx);
   }
 
+  /*  
+  cudnn requires tensor dimensions to fulfill 2 requirements:
+    1.) dimensions to be set to a minimum of 4 for 4d and lower dimensional tensors (cudnnOp supports up to 5d, cudnnReduce supports up to 8d)
+    2.) dimensions have to be set to multiples of 8
+
+    for e.g. Tensor A has shape {3,3}, cudnn requires shape of {1,1,24,24} to be the input
+             Tensor B has shape (2,3,4), cudnn requires shape of {1,16,24,32} to be the input
+  */
+  vector<int> generate_shape_cuda() const {
+    if(shape_.size() <= 4){
+      vector<int> shape_arr = {1,1,1,1};
+      for (size_t n=0; n<shape_.size(); ++n) {
+        shape_arr[4 - shape_.size() + n] = shape_.at(n)*8;
+      }
+      return shape_arr;
+    } else if(shape_.size() == 5){
+      vector<int> shape_arr = {1,1,1,1,1};
+      for (int n=0; n<5; ++n) {
+        shape_arr[n] = shape_.at(n)*8;
+      }
+      return shape_arr;
+    } else {
+      LOG(FATAL) << "Dimensions (shape) beyond 5 are currently not supported" ;
+    }
+  }
+
+  int generate_dim_cuda() const {
+    if(shape_.size() <= 4){return 4;}
+    else if(shape_.size() == 5){return 5;}
+    else{
+      LOG(FATAL) << "Dimensions (shape) beyond 5 are currently not supported" ;
+    } 
+  }
+
   size_t nDim() const { return shape_.size(); }
 
   bool empty() const { return nDim() == 0; }
 
-  bool transpose() const { return transpose_; }
+  //bool transpose() const { return transpose_; }
+  bool transpose() const { return (strides_.back() != 1); }
+
+  const vector<int>& strides() const { return strides_; }
+
+  /*  
+  cudnn requires stride dimensions to conform to the format of the shape input as well
+    1.) stride dimensions to be set to a minimum of 4 for 4d and lower dimensional tensors (cudnnOp supports up to 5d, cudnnReduce supports up to 8d)
+    2.) stride dimensions have to be set to powers of 8, depending on the stride order (outer stride = higher power)
```

Comment:
I still can't find the documentation about this.
Only some special operations require the tensors to be aligned. 
Most Tensors do not need to be aligned.

Reply:
I am not too sure as well, cudnn documentations are kind of scarce online and there are very little questions/answers related to cudnn on stackoverflow as well.

I figured these out through trials and errors. So far, this is the only format that generates the correct output for various functions such as add, op, scale and transform. For reduce, only the sum function is working (e.g. by reducing all axes) but not the RowMax function (e.g. reducing specific or axis=1). Currently, RowMax is implemented using transform and cuda instead.